src/host/screenInfo.cpp

```diff
@@ -1239,23 +1242,21 @@ void SCREEN_INFORMATION::_InternalSetViewportSize(const COORD* const pcoordSize,
         srNewViewport.Top = std::max<SHORT>(0, srNewViewport.Top - offBottomDelta);
     }
 
-    // See MSFT:19917443
-    // If we're in terminal scrolling mode, and we've changed the height of the
-    //      viewport, the new viewport's bottom to the _virtualBottom.
-    // GH#1206 - Only do this if the viewport is _growing_ in height. This can
-    // cause unexpected behavior if we try to anchor the _virtualBottom to a
-    // position that will be greater than the height of the buffer.
-    const auto& gci = ServiceLocator::LocateGlobals().getConsoleInformation();
-    auto newViewport = Viewport::FromInclusive(srNewViewport);
-    if (gci.IsTerminalScrolling() && newViewport.Height() >= _viewport.Height())
+    // In general we want to avoid moving the virtual bottom unless it's aligned with
+    // the visible viewport, so we check whether the changes we're making would cause
+    // the bottom of the visible viewport to intersect the virtual bottom at any point.
+    // If so, we update the virtual bottom to match. We also update the virtual bottom
+    // if it's less than the new viewport height minus 1, because that would otherwise
+    // leave the virtual viewport extended past the top of the buffer.
+    const auto newViewport = Viewport::FromInclusive(srNewViewport);
+    if ((_virtualBottom >= _viewport.BottomInclusive() && _virtualBottom < newViewport.BottomInclusive()) ||
+        (_virtualBottom <= _viewport.BottomInclusive() && _virtualBottom > newViewport.BottomInclusive()) ||

```

Comment:
I had a bit trouble understanding this visually until I realized that it basically moves the virtual bottom if the viewport changed the side it is relative to the virtual bottom. In other words: If the viewport used to be above the virtual bottom and is now below, then the virtual bottom is moved and vice versa.

In order to understand this I expressed the boolean condition in an equivalent "numeric" expression. These two lines are equivalent to:
```cpp
const auto delta1 = _virtualBottom - _viewport.BottomInclusive();
const auto delta2 = _virtualBottom - newViewport.BottomInclusive();
// Notes:
// * `delta2 != 0` is required because _virtualBottom and newViewport.BottomInclusive()
//   are only ever compared via < and >
// * `delta1 == 0` is required because _virtualBottom and _viewport.BottomInclusive()
//    are also implicitly compared for == (due to >= / <=)
// * (a ^ b) < 0 is true if a/b have opposing sign bits
delta2 != 0 && (delta1 == 0 || (delta1 ^ delta2) < 0);
```

It seems it doesn't matter whether we check for `delta2 == 0`, because if it is 0, then `_virtualBottom == srNewViewport.Bottom` and the line below won't change anything. This is why we can also write:
```cpp
const auto delta1 = _virtualBottom - _viewport.BottomInclusive();
const auto delta2 = _virtualBottom - newViewport.BottomInclusive();
delta1 == 0 || (delta1 ^ delta2) < 0;
```

With this expression in place I found it easier to understand that the virtual bottom moves if either:
* the virtual bottom is currently following the viewport along (`delta1 == 0`) 
* the new viewport has changed direction and is now on the other side
  relative to the virtual bottom (`(delta1 ^ delta2) < 0`)

Reply:
> I had a bit trouble understanding this visually until I realized that it basically moves the virtual bottom if the viewport changed the side it is relative to the virtual bottom. In other words: If the viewport used to be above the virtual bottom and is now below, then the virtual bottom is moved and vice versa.

The code as written is almost an exact reproduction of your final sentence. I don't get how you can find that difficult to understand? It's essentially self-documenting.

Whereas this looks to me like the winning entry in a code-obfuscation contest!

    delta1 == 0 || (delta1 ^ delta2) < 0;

If you were suggesting that as a performance improvement, I'd get it. But I can't fathom how you see that as easier to understand? Our minds clearly work in completely different ways. But if someone else chimes in here agreeing with you, I'm happy to be overruled.

Although I should also ask, how sure are you about the portability of that code? Is it really safe to XOR negative numbers like that?

Reply:
> The code as written is almost an exact reproduction of your final sentence.

No, not as I perceive it. You wrote:
> [...] we check whether the changes we're making would cause the bottom of the visible viewport to intersect the virtual bottom at any point.

So I was wondering, how would you know that the visible viewport is intersecting the virtual bottom by just comparing `BottomInclusive()` but not `Top()`? Because your code is disagreeing with your comment: It's not intersecting a rectangle ("viewport") with a line. It's comparing the position of 2 lines (old and new viewport bottom) relative to 1 "point of reference" line (the current `_virtualBottom`). The condition is true if the 2 lines are on opposing sides of the reference.

> But I can't fathom how you see that as easier to understand? Our minds clearly work in completely different ways.

I'm sorry. 😥
I think I could properly explain it in person, but I'll try it via text anyways.

I read your explanation and then the first half of the first line of your if condition - `_virtualBottom >= _viewport.BottomInclusive()`. Afterwards I tried to join that with the second half `_virtualBottom < newViewport.BottomInclusive()` and I had about this image in my mind:
<img width="458" alt="12972_viewport" src="https://user-images.githubusercontent.com/2256941/165757074-47515c21-db85-4f6b-9877-945b6e03412f.png">

I mean clearly the bottom of the `newViewport` must be below the `_virtualBottom`, but where? What does it mean to the terminal if a viewport's bottom "intersects" a line? And most importantly: What happens if the `_viewport` is below the `_virtualBottom`?

In either case I guess what I'm trying to say is that it took me a bit to grok that the code is basically just:
```
   (oldViewport <= virtualBottom < newViewport)
|| (oldViewport >= virtualBottom > newViewport)
```
...but the `<` and `<=` sides are flipped in your code.

In human terms one could say: "Jiggle the virtual bottom around as long as it's in between the bottoms of the old and new viewport."

And I find this "in between old and new" just as easy to reason about if one were to write `differentSignBit(ref - new, ref - old)` or equivalent `(ref < new) != (ref < old)` (which btw also works for your if condition). The `differentSignBit` thing makes sense to me, because the difference `ref - new` and `ref - old` are only ever negative if the new/old value is smaller than the reference point. Since we want them on opposing sides, we're basically asking them to have different signs. I... I just like that. 😥

But even at this point there was one "aha" realization left about why the `newViewport` is compared with `<` and the old one with `<=`. Of course it was easy to see then that this distinction doesn't matter, because they lead to the same outcome.

Basically, to finish my ramblings and leave it at something useful... This is the same code, but all comparison signs are the same and point in the same direction:
```cpp
if ((_viewport.BottomInclusive() <= _virtualBottom && _virtualBottom <= newViewport.BottomInclusive()) ||
    (newViewport.BottomInclusive() <= _virtualBottom && _virtualBottom <= _viewport.BottomInclusive()) ||
```

And now we _could_ abstract that as:
```cpp
static constexpr auto isBetween = [](auto lo, auto mid, auto hi) {
    return (lo <= mid) && (mid <= hi);
}

if (isBetween(_viewport.BottomInclusive(), _virtualBottom, newViewport.BottomInclusive()) ||
    isBetween(newViewport.BottomInclusive(), _virtualBottom, _viewport.BottomInclusive()) ||
```

> Although I should also ask, how sure are you about the portability of that code? Is it really safe to XOR negative numbers like that?

Yeah, I'm fairly certain. Bitwise operations on signed integers are defined to operate on their bitwise representation and since both integers are of the same type, the (undefined) position of their sign bits are the same, making this operation well defined.
But this doesn't matter as much anymore, because C++20 requires signed integers to use two's complement without padding bits. 🙂

Reply:
> So I was wondering, how would you know that the visible viewport is intersecting the virtual bottom by just comparing `BottomInclusive()` but not `Top()`?

You missed part of the comment - I said "the _bottom_ of the visible viewport". The top of the viewport is not relevant at all.

> But even at this point there was one "aha" realization left about why the `newViewport` is compared with `<` and the old one with `<=`. Of course it was easy to see then that this distinction doesn't matter, because they lead to the same outcome.

OK, I can see how that might be confusing, and I'd be happy to make both include the `=` if that's clearer. If it's already equal to the new viewport bottom it's essentially a noop, but that doesn't really matter.

> Basically, to finish my ramblings and leave it at something useful... This is the same code, but all comparison signs are the same and point in the same direction:
> 
> ```c++
> if ((_viewport.BottomInclusive() <= _virtualBottom && _virtualBottom <= newViewport.BottomInclusive()) ||
>     (newViewport.BottomInclusive() <= _virtualBottom && _virtualBottom <= _viewport.BottomInclusive()) |
> ```

I did consider ordering it that way, but in the end I thought it might be clearer to align it with how it was described in the comments. I don't think it makes much difference either way, though, so again I'd be happy to make that change, if that's what you're proposing.

What I don't get is how you reached that point, and then decided it would be even easier to understand if you rewrote everything with a bitwise XOR operator.

> C++20 requires signed integers to use two's complement without padding bits.

That's nice to know. I never would expected that.

Reply:
> What I don't get is how you reached that point, and then decided it would be even easier to understand if you rewrote everything with a bitwise XOR operator.

You know, I think I got it the first time.
I'll make sure to not comment something like that again.

Reply:
Sorry, I didn't meant to be overly critical of that suggestion. I'm just trying to figure out what exactly I should be changing this code to so it's understandable for everyone. In short, would you be happy with the reordered condition with all the comparisons the same?